### PR TITLE
Parallel env for policy runner eval

### PR DESCRIPTION
**Pull request by xyao-nv**
Enable parallel env evaluation for gr00t policy.

- Add `env_ids` in `reset()` args of `policy_base.py`
- Expand 1st dim in gr00t action chunk tensor as num_envs
- Add policy reset for action chunk from terminated env_id only
- Re-able multi-envs tests

### CODE REVIEW

File: isaaclab_arena_gr00t/gr00t_closedloop_policy.py [hunk 1]
```diff
@@ -45,6 +48,21 @@ def __init__(self, policy_config_yaml_path: Path, num_envs: int = 1, device: str
...
+            # GR00T outputs are used for WBC inputs dim. So adding WBC commands to the action dim.
+            # WBC commands: navigate_command, base_height_command, torso_orientation_rpy_command
+            self.action_dim += NUM_NAVIGATE_CMD + NUM_BASE_HEIGHT_CMD + NUM_TORSO_ORIENTATION_RPY_CMD
+
+        # initialize to NaN to indicate that the action chunk is not yet computed
+        self.current_action_chunk = torch.ones(
+            (num_envs, self.num_feedback_actions, self.action_dim),
+            dtype=torch.float32,
+            device=device,
+        ) * float("nan")
```
Comment (alexmillane) [inline isaaclab_arena_gr00t/gr00t_closedloop_policy.py]: Suggestion to switch from using Nan as a sentinal to just having a boolean vector which indicates a specific env needs a reset. Then we can avoid all of this code for setting and detecting nans. Nans are kind of dangerous, so ideally we'd not introduce them.

Do you see any issue with just having a list `env_requires_new_action_chunk = [0] * self.num_envs`?

File: isaaclab_arena_gr00t/gr00t_closedloop_policy.py [hunk 2]
```diff
@@ -118,23 +136,45 @@ def get_observations(self, observation: dict[str, Any], camera_name: str = "robo
...
+        # assert for all env_ids that two conditions are met:
+        # 1. the action chunk is not NaN
+        # 2. -1 < the action index < num_feedback_actions
+        assert (
+            not torch.isnan(self.current_action_chunk).any(dim=1).any(dim=1).any()
+        ), "At least one env's action chunk is NaN"
+        assert (
+            self.current_action_index.max() < self.num_feedback_actions
+        ), "At least one env's action index is greater than the number of feedback actions"
+        assert self.current_action_index.min() >= 0, "At least one env's action index is less than 0"
```
Comment (alexmillane) [inline isaaclab_arena_gr00t/gr00t_closedloop_policy.py]: Nice sanity checking!

File: isaaclab_arena_gr00t/gr00t_closedloop_policy.py [hunk 3]
```diff
@@ -45,6 +48,21 @@ def __init__(self, policy_config_yaml_path: Path, num_envs: int = 1, device: str
...
 
+        self.action_dim = len(self.robot_action_joints_config)
+        if self.task_mode == TaskMode.G1_LOCOMANIPULATION:
+            # GR00T outputs are used for WBC inputs dim. So adding WBC commands to the action dim.
+            # WBC commands: navigate_command, base_height_command, torso_orientation_rpy_command
+            self.action_dim += NUM_NAVIGATE_CMD + NUM_BASE_HEIGHT_CMD + NUM_TORSO_ORIENTATION_RPY_CMD
+
+        # initialize to NaN to indicate that the action chunk is not yet computed
+        self.current_action_chunk = torch.ones(
+            (num_envs, self.num_feedback_actions, self.action_dim),
```
Comment (alexmillane) [inline isaaclab_arena_gr00t/gr00t_closedloop_policy.py]: what do you think about renaming `self.num_feedback_actions` to `self.action_chunk_length`?

File: isaaclab_arena_gr00t/gr00t_closedloop_policy.py [hunk 4]
```diff
@@ -118,23 +136,45 @@ def get_observations(self, observation: dict[str, Any], camera_name: str = "robo
...
-        if self.current_action_chunk is None and self.current_action_index == 0:
-            self.current_action_chunk = self.get_action_chunk(observation, self.policy_config.pov_cam_name_sim)
-            assert self.current_action_chunk.shape[1] >= self.num_feedback_actions
-
-        assert self.current_action_chunk is not None
-        assert self.current_action_index < self.num_feedback_actions
-
-        action = self.current_action_chunk[:, self.current_action_index]
-        assert action.shape == env.action_space.shape, f"{action.shape=} != {env.action_space.shape=}"
+        """Get the action from the current action chunk. If action chunk is not valid, compute a new action chunk."""
```
Comment (alexmillane) [inline isaaclab_arena_gr00t/gr00t_closedloop_policy.py]: suggestion to add a doc string here. Mainly to get the dimensions of the output. With all the dimensions flying around, I feel that this will help a user.

File: isaaclab_arena_gr00t/gr00t_closedloop_policy.py [hunk 5]
```diff
@@ -160,12 +200,19 @@ def get_action_chunk(self, observation: dict[str, Any], camera_name: str = "robo
         elif self.task_mode == TaskMode.GR1_TABLETOP_MANIPULATION:
             action_tensor = robot_action_sim.get_joints_pos()
 
-        assert action_tensor.shape[1] >= self.num_feedback_actions
+        assert action_tensor.shape[0] == self.num_envs and action_tensor.shape[1] >= self.num_feedback_actions
         return action_tensor
 
-    def reset(self):
+    def reset(self, env_ids: torch.Tensor | None = None):
```
Comment (alexmillane) [inline isaaclab_arena_gr00t/gr00t_closedloop_policy.py:219]: Suggestion to add a docstring now that the reset behaviour is a bit complicated.

File: isaaclab_arena_gr00t/gr00t_closedloop_policy.py [hunk 6]
```diff
@@ -118,23 +136,45 @@ def get_observations(self, observation: dict[str, Any], camera_name: str = "robo
...
-            self.current_action_index = 0
+
+        # for those rows in current_action_chunk that equal to num_feedback_actions, reset to NaN
+        reset_env_ids = self.current_action_index == self.num_feedback_actions
+        self.current_action_chunk[reset_env_ids] = 1 * float("nan")
+        # set the action index for those env_ids to -1 to indicate that the action chunk is reset
+        self.current_action_index[reset_env_ids] = -1
         return action
 
     def get_action_chunk(self, observation: dict[str, Any], camera_name: str = "robot_head_cam_rgb") -> torch.Tensor:
```
Comment (alexmillane) [inline isaaclab_arena_gr00t/gr00t_closedloop_policy.py:182]: Suggestion to add a docstring with the output dimensions as things are getting complicated.

File: isaaclab_arena_gr00t/gr00t_closedloop_policy.py [hunk 7]
```diff
@@ -160,12 +200,19 @@ def get_action_chunk(self, observation: dict[str, Any], camera_name: str = "robo
...
         return action_tensor
 
-    def reset(self):
+    def reset(self, env_ids: torch.Tensor | None = None):
         """Resets the policy's internal state."""
         # As GR00T is a single-shot policy, we don't need to reset its internal state
         # Only reset the action chunking mechanism
-        self.current_action_chunk = None
-        self.current_action_index = 0
+        print(f"resetting action chunk for env_ids: {env_ids}")
```
Comment (alexmillane) [inline isaaclab_arena_gr00t/gr00t_closedloop_policy.py]: suggestion to move this printing out to the closed loop script

File: isaaclab_arena_gr00t/gr00t_closedloop_policy.py (hunk 3 above)
Comment (xyao-nv) [reply isaaclab_arena_gr00t/gr00t_closedloop_policy.py]: good point as the use of action chunk is explicit now.

File: isaaclab_arena_gr00t/gr00t_closedloop_policy.py (hunk 1 above)
Comment (xyao-nv) [reply isaaclab_arena_gr00t/gr00t_closedloop_policy.py]: Great suggestion as it does not introduce too much memory.

File: isaaclab_arena_gr00t/gr00t_closedloop_policy.py (hunk 4 above)
Comment (xyao-nv) [reply isaaclab_arena_gr00t/gr00t_closedloop_policy.py]: done

File: isaaclab_arena_gr00t/gr00t_closedloop_policy.py (hunk 6 above)
Comment (xyao-nv) [reply isaaclab_arena_gr00t/gr00t_closedloop_policy.py:182]: done

File: isaaclab_arena_gr00t/gr00t_closedloop_policy.py (hunk 5 above)
Comment (xyao-nv) [reply isaaclab_arena_gr00t/gr00t_closedloop_policy.py:219]: done

File: isaaclab_arena_gr00t/gr00t_closedloop_policy.py (hunk 7 above)
Comment (xyao-nv) [reply isaaclab_arena_gr00t/gr00t_closedloop_policy.py]: done
